### ISSUE TITLE
Buffer copies

### ISSUE THREAD

**Issue by dcousens** [varstruct/varstruct#33]
If a `Buffer` is to be returned,  we return a copy.

Why?

In most cases I think people assume `Buffer` is only copied when necessary,  and if they intend to mutate,  I'd wager they copy it themselves.

Can we remove the extra `Buffer.from`s/

**Comment by fanatid** [comment]
This is definitely not a bug :)
to be honest not sure what will better -- return copy or not

**Comment by dominictarr** [comment]
copying is generally safer, unless it's a secret value and you want to be sure it's zeroed later.
copying can be a performance issue, but it's fairly hard to find a case where you really need performance that much (let alone, will be able to measure the difference it makes, or not have something else that is more of the bottleneck)

certainly, it's not worth the effort unless you have a specific case that really needs it, in which case you should start with some benchmarks, etc.

in absense of a specific case that really needs it, I'd go with copy.

**Comment by dcousens** [comment]
@dominictarr well, as always, I'm profiling/optimizing,  not speculating,  and yeah,  `varstruct` takes up 45% of my application loading time when loading about 120kB of complex data structures.... and copying of `Buffer`s takes up about 20-25% of that.

I'm trying to see if I can get it within a magnitude of `JSON.parse`, which is comparatively 30x faster at the moment.

The biggest hit we already nuked with https://github.com/varstruct/varstruct/pull/30.

**Comment by dominictarr** [comment]
@dcousens it that case, I support returning! but you should tell us something about your application! since secure-scuttlebutt also has some in memory structures I've been thinking about in buffer based memory structures  - one approach would be to have stuff you can just read and use in-place (no parsing!) but if you have to have hashtables, or complex trees, and GC your own memory, it might get pretty hard to beat V8.

**Comment by dcousens** [comment]
Indeed,  I have thought about lazy `decode` as an option,  which we could easily support if each child of an object has a fixed `encodingLength`.

**Comment by dcousens** [comment]
For the record,  I'm not trying to beat `JSON.parse`,  just,  less than 10x performance loss.

**Comment by dcousens** [comment]
@fanatid any final thoughts here?

**Comment by fanatid** [comment]
I think that copy is better

**Comment by dcousens** [comment]
@fanatid the performance is abysmal though?

**Comment by ralphtheninja** [comment]
@dcousens New module `varstrict` 😄

**Comment by dcousens** [comment]
@fanatid if the concern is dev experience,  and copies protect the dev,  I agree.

Maybe we could detect if the call is recursive to avoid copies of copies of copies...
